File: library/analytics/src/main/java/org/cru/godtools/analytics/model/AnalyticsActionEvent.java
package org.cru.godtools.analytics.model;

import android.net.Uri;
import android.text.TextUtils;

import java.util.Locale;
import java.util.Map;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class AnalyticsActionEvent extends AnalyticsBaseEvent {
    private static final String SNOWPLOW_CONTENT_SCORING_URI_PATH_ACTION = "action";

    @Nullable
    private final String mCategory;
    @NonNull
    private final String mAction;
    @Nullable
    private final String mLabel;

    protected AnalyticsActionEvent() {
        this(null, "", null, null);
    }

    public AnalyticsActionEvent(@Nullable final String category, @NonNull final String action) {
        this(category, action, null, null);
    }

    public AnalyticsActionEvent(@Nullable final String category, @NonNull final String action,
                                @Nullable final String label) {
        this(category, action, label, null);
    }

    public AnalyticsActionEvent(@Nullable final String category, @NonNull final String action,
                                @Nullable final String label, @Nullable final Locale locale) {
        super(locale);
        mCategory = category;
        mAction = action;
        mLabel = label;
    }

    @Nullable
    public String getCategory() {
        return mCategory;
    }

    @NonNull
    public String getAction() {
        return mAction;
    }

    @Nullable
    public String getLabel() {
        return mLabel;
    }

    @Nullable
    public Map<String, ?> getAttributes() {
        return null;
    }

    @NonNull
    @Override
    public Uri.Builder getSnowPlowContentScoringUri() {
        return super.getSnowPlowContentScoringUri()
                .authority(SNOWPLOW_CONTENT_SCORING_URI_PATH_ACTION)
                .appendPath(getCategory())
                .appendPath(getAction());
    }

    @Override
    public String getSnowPlowPageTitle() {
        StringBuilder stringBuilder = new StringBuilder();

        if (!TextUtils.isEmpty(getCategory())) {
            stringBuilder.append(getCategory()).append(" : ");
        }
        stringBuilder.append(getAction());

        if (!TextUtils.isEmpty(getLabel())) {
            stringBuilder.append(" : ").append(getLabel());
        }
        return stringBuilder.toString();
    }
}

File: library/analytics/src/main/kotlin/org/cru/godtools/analytics/model/AnalyticsActionEvent.kt
package org.cru.godtools.analytics.model

import android.os.Bundle
import java.util.Locale
import javax.annotation.concurrent.Immutable

@Immutable
open class AnalyticsActionEvent(
    val action: String,
    val label: String? = null,
    locale: Locale? = null,
    systems: Set<AnalyticsSystem> = DEFAULT_SYSTEMS
) : AnalyticsBaseEvent(locale, systems) {
    constructor(action: String, label: String? = null, locale: Locale? = null, system: AnalyticsSystem) :
        this(action, label, locale, setOf(system))

    open val firebaseEventName get() = action
    open val firebaseParams get() = Bundle()

    override fun equals(other: Any?) = when {
        this === other -> true
        javaClass != other?.javaClass -> false
        !super.equals(other) -> false
        other !is AnalyticsActionEvent -> false
        action != other.action -> false
        label != other.label -> false
        else -> true
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + action.hashCode()
        result = 31 * result + (label?.hashCode() ?: 0)
        return result
    }
}
